wallet-new/src/Cardano/Wallet/Kernel/BListener.hs

```diff
@@ -95,34 +163,191 @@ applyBlock pw@PassiveWallet{..} b = do
+                      Just moreErrs -> handleApplyBlockErrors k moreErrs -- Try again, better luck next time.
+
+      -- Try to apply a single block, failing if it does not fit onto the most recent checkpoint.
+      applyOneBlock :: Node.SecurityParameter
+                    -> Maybe (Set HdAccountId)
+                    -> ResolvedBlock
+                    -> ExceptT (NonEmptyMap HdAccountId ApplyBlockFailed) IO ()
+      applyOneBlock k accts b' = ExceptT $ do
+          ((ctxt, blocksByAccount), metas) <- prefilterBlock' pw b'
+          -- apply block to all Accounts in all Wallets
+          mConfirmed <- update' _wallets $ ApplyBlock k ctxt accts blocksByAccount
+          case mConfirmed of
+              Left  errs      -> return (Left errs)
+              Right confirmed -> do
+                  modifyMVar_ _walletSubmission (return . Submission.remPending confirmed)
+                  mapM_ (putTxMeta _walletMeta) metas
+                  createCheckpointIfNeeded
+                  return $ Right ()
+
+      -- Determine if a failure in 'ApplyBlock' was due to the account being ahead, behind,
+      -- or incomparable with the provided block.
+      classifyFailure :: ApplyBlockFailed -> IO ApplyBlockErrorCase
+      classifyFailure (ApplyBlockNotSuccessor curCtx cpCtx) = do
+          result <- runExceptT (findMissing (Just curCtx) cpCtx [])
+          case result of
+              Right blocks -> return (AccountIsBehindBlock blocks)
+              Left (CouldNotReachCheckpoint _)    ->
+                  -- Figure out if the checkpoint is incomparable, or from the future.
+                  runExceptT (findMissing cpCtx (Just curCtx) []) <&> \case
+                      Right _ -> AccountIsAheadOfBlock
+                      Left  _ -> AccountIsOnWrongFork
+              Left (CouldNotFindBlockForHeader _) -> return AccountIsOnWrongFork
+              Left (NotAMainBlock _)              -> return AccountIsOnWrongFork
+              Left (SuccessorChanged _ _)         -> return AccountIsOnWrongFork
+
+      -- Find all blocks that were missing between the given block and the wallet's most recent
+      -- checkpoint. 'Nothing' is used to represent the genesis block.
+      findMissing :: Maybe BlockContext
+                  -> Maybe BlockContext
+                  -> [ResolvedBlock]
+                  -> ExceptT BackfillFailed IO (OldestFirst [] ResolvedBlock)
+      findMissing Nothing    Nothing    acc = return (OldestFirst acc)
+      findMissing Nothing    (Just cp) _acc = throwError (CouldNotReachCheckpoint cp)
+      findMissing (Just cur) tgt        acc =
+        if (Just (cur ^. bcHash)) == (tgt ^? _Just . bcHash) then
+            return (OldestFirst acc)
+        else do
+            rb   <- hashToBlock (cur ^. bcHash . fromDb)
+            prev <- traverse hashToBlock (rb ^? rbContext . bcPrevMain . _Just . fromDb)
+            rb `seq` findMissing (prev ^? _Just . rbContext) tgt (rb : acc)
```

Comment:
Interesting -- I am surprised you decided to force `rb` in whnf rather than `acc`. My review point was that we never force `acc` to be evaluated till the last minute and this is problematic as every call to `:` would create a thunk and that can leak memory.

Am I being totally undercaffinated here? Seems like this change makes the values of the data structure strict rather than the spine.

Reply:
Oh, I thought that's what you were proposing. So for your proposal, it would be sufficient to change the `acc` parameters to `!acc`, yes?

Reply:
yup, exactly!